### PR TITLE
add auth v3

### PR DESCRIPTION
**Pull request by DoubleSpout**
1.  add etcdv3 auth (support etcd 3.3+)
2.  update v2 and v3 cluster testcase
3. fix v3 cluster host bug

### CODE REVIEW

File: lib/resty/etcd/v3.lua [hunk 1]
```diff
@@ -160,6 +190,37 @@ local function choose_endpoint(self)
     return endpoints[pos]
 end
 
+-- return refresh_is_ok, error
+refresh_jwt_token = function (self)
```
Comment (membphis) [inline lib/resty/etcd/v3.lua]: `refresh_jwt_token = function (self)` 
change to:
 `function refresh_jwt_token(self)`

File: t/v2/add-auth.sh [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+curl -X "PUT" "${AUTH_ENDPOINT_V2}/v2/auth/users/root" -H "Content-Type: application/json; charset=utf-8" -d $"{\"user\":\"${AUTH_USER}\",\"password\":\"${AUTH_PWD}\"}"
+curl -X "PUT" "${AUTH_ENDPOINT_V2}/v2/auth/enable"
+curl -X "PUT" "${AUTH_ENDPOINT_V2}/v2/auth/roles/guest" -H "Content-Type: text/plain; charset=utf-8" -u "${AUTH_USER}:${AUTH_PWD}" -d $'{"role":"guest","revoke":{"kv":{"write": ["/*"]}}}'
+
```
Comment (membphis) [inline t/v2/add-auth.sh]: remove one blank line

File: t/v2/cluster.t [hunk 3]
```diff
@@ -81,24 +81,28 @@ checked error msg as expect: Key not found
 all done
 
 
-
-=== TEST 2: cluster set + delete + get
+=== TEST 2: cluster set + delete + get + auth
```
Comment (membphis) [inline t/v2/cluster.t]: We'd better add a new test case for testing new case.

File: t/v3/add-auth.sh [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+export ETCDCTL_API=3 
+etcdctl --endpoints="${AUTH_ENDPOINT_V3}" user add "${AUTH_USER}:${AUTH_PWD}"
+etcdctl --endpoints="${AUTH_ENDPOINT_V3}" role add root
+etcdctl --endpoints="${AUTH_ENDPOINT_V3}" user grant-role root root
+etcdctl --endpoints="${AUTH_ENDPOINT_V3}" role list
+etcdctl --endpoints="${AUTH_ENDPOINT_V3}" user list
+etcdctl --endpoints="${AUTH_ENDPOINT_V3}" auth enable
```
Comment (membphis) [inline t/v3/add-auth.sh]: Need a blank line at the end

File: lib/resty/etcd/v3.lua [hunk 5]
```diff
@@ -446,8 +519,10 @@ local function request_chunk(self, method, host, port, path, opts, timeout)
 
         if body.result.events then
             for _, event in ipairs(body.result.events) do
-                event.kv.value = decode_base64(event.kv.value)
-                event.kv.value = decode_json(event.kv.value)
+                if event.kv.value then   -- DELETE not have value
```
Comment (membphis) [inline lib/resty/etcd/v3.lua]: we can do this at a separate PR, one PR only do one thing. It will make the PR smaller and easier to review.

File: lib/resty/etcd/v3.lua (hunk 1 above)
Comment (DoubleSpout) [reply lib/resty/etcd/v3.lua]: fixed

File: t/v2/cluster.t (hunk 3 above)
Comment (DoubleSpout) [reply t/v2/cluster.t]: cluster case is mian test auth

File: t/v3/add-auth.sh (hunk 4 above)
Comment (DoubleSpout) [reply t/v3/add-auth.sh]: fix

File: lib/resty/etcd/v3.lua (hunk 5 above)
Comment (DoubleSpout) [reply lib/resty/etcd/v3.lua]: this is last pr

File: .travis.yml [hunk 6]
```diff
@@ -36,7 +36,7 @@ env:
 
 install:
   - git clone https://github.com/openresty/test-nginx.git test-nginx
-  - sudo luarocks install lua-resty-http
+  - sudo luarocks install lua-resty-http 0.14-0
```
Comment (membphis) [inline .travis.yml]: Why do we have to use the 14.0 version?

Comment (DoubleSpout) [reply .travis.yml]: I got a `unread buffer` exception from `lua-resty-http`, so I try to use last release version `0.14.0`,finally I found the `keep-alive` params to fix it.

Comment (DoubleSpout) [reply .travis.yml]: I got a `unread buffer` exception from `lua-resty-http` in etcd v3.4.0, so I try to use last release version `0.14.0`,finally I found the `keep-alive` params to fix it.

Comment (membphis) [reply .travis.yml]: https://github.com/iresty/lua-resty-etcd/blob/master/rockspec/lua-resty-etcd-master-0.1-0.rockspec#L16

I think we should keep same with this.

Comment (DoubleSpout) [reply .travis.yml]: https://github.com/iresty/lua-resty-etcd/pull/41/commits/3f0dc6399856b0a7fa0c7ded2353f08638995c5a

modify back to default

File: .travis.yml [hunk 7]
```diff
@@ -36,7 +36,7 @@ env:
 
 install:
   - git clone https://github.com/openresty/test-nginx.git test-nginx
-  - sudo luarocks install lua-resty-http 0.14-0
+  - sudo luarocks install lua-resty-http
   - sudo luarocks install lua-typeof
```
Comment (membphis) [inline .travis.yml]: BWT, I think we can use `sudo make dev` to install dependencies via Luarocks.

We can do this in a new PR later.

File: lib/resty/etcd/v3.lua [hunk 8]
```diff
@@ -160,6 +194,37 @@ local function choose_endpoint(self)
...
+                        opts, 5, true)    -- default authenticate timeout 5 second
+    if err then
+        return nil, err
+    end
+
+    if not res or not res.body or not res.body.token then
+        return nil, 'authenticate refresh token fail'
+    end
+
+    self.jwt_token = res.body.token
```
Comment (membphis) [inline lib/resty/etcd/v3.lua]: When we access an etcd cluster, can this JWT_TOKEN be used on different etcd nodes?

**Comment by membphis** [comment]
code style:

```shell
$ lj-releng
./lib/resty/etcd.lua: 0.9 (0.9)
Checking line length exceeding 80...
Checking line length exceeding 80...
./lib/resty/etcd/v2.lua: host
Checking line length exceeding 80...
Checking line length exceeding 80...
./lib/resty/etcd/v3.lua:213:                        choose_endpoint(self).full_prefix .. "/auth/authenticate",
./lib/resty/etcd/v3.lua:214:                        opts, 5, true)    -- default authenticate timeout 5 second
```

File: lib/resty/etcd/v3.lua (hunk 8 above)
Comment (DoubleSpout) [reply lib/resty/etcd/v3.lua]: yes the etcd cluster will sync the auth

**Comment by membphis** [comment]
@DoubleSpout many thanks, merged already

File: lib/resty/etcd/v3.lua [hunk 9]
```diff
@@ -135,10 +164,15 @@ function _M.new(opts)
     end
 
     return setmetatable({
-            timeout   = timeout,
-            ttl       = ttl,
-            is_cluster= #endpoints > 1,
-            endpoints = endpoints,
+            last_auth_time = now(), -- save last Authentication time
+            jwt_token   = nil,       -- last Authentication token
+            is_auth     = not not (user and password),
```
Comment (membphis) [inline lib/resty/etcd/v3.lua]: `not not ...` why??

**Comment by Yiyiyimu** [comment]
Just for note, auth in etcd grpc-gateway is supported since v3.3, as discussed in [this issue](https://github.com/etcd-io/etcd/issues/6643).

**Comment by membphis** [comment]
> Just for note, auth in etcd grpc-gateway is supported since v3.3, as discussed in [this issue](https://github.com/etcd-io/etcd/issues/6643).

we can write this to API doc, please make a check.

**Comment by Yiyiyimu** [comment]
> we can write this to API doc, please make a check.

Sure.
